File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>>ans;
    void combination(vector<int>& nums,int tar, int i,vector<int>& v){
        if(tar==0){
            ans.push_back(v);
            return;
        }
        for(int j=i;j<nums.size();j++){
            if(j>i && nums[j]==nums[j-1]) continue;
            if(nums[j]>tar) break;
            v.push_back(nums[j]);
            combination(nums,tar-nums[j],j+1,v);
            v.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<int>v;
        combination(candidates,target,0,v);
        
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {

    public void helper(int[] cand,int tar,int start, List<List<Integer>> ans,List<Integer> temp){

        if(tar==0){
            ans.add(new ArrayList(temp));
            return;
        }

        if (tar < 0) return;

        for (int i = start; i < cand.length; i++) {
            if (i > start && cand[i] == cand[i - 1]) continue; 

            temp.add(cand[i]);
            helper(cand, tar - cand[i], i + 1, ans, temp);
            temp.remove(temp.size() - 1); 
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> temp=new ArrayList<>();
        Arrays.sort(candidates);
        helper(candidates,target,0,ans,temp);
        return ans;
    }
}